packages/engine/tests/engine/input/systems/ClientInputSystem.test.tsx

```diff
@@ -0,0 +1,292 @@
+    globalThis.window.addEventListener = mockWinEvents.addEventListener as any
+    globalThis.window.removeEventListener = mockWinEvents.removeEventListener as any
+    globalThis.navigator = new MockNavigator() as any
+
+    const cleanup = addClientInputListeners()
+
+    assert(typeof cleanup === 'function')
+
+    const mock: MockEngineRenderer = EngineRenderer.instance as MockEngineRenderer
+    const mockDomElm = mock.renderer.domElement as unknown as MockEventListener
+    assert(mockDomElm.listeners.length > 1, 'Callbacks were added to canvas')
+    mockDomElm.listeners.forEach((listener) => {
+      listener(mockEvent)
+    })
+
+    assert(mockDocEvents.listeners.length > 1, 'Callbacks were added to document')
+    mockDocEvents.listeners.forEach((listener) => {
+      listener(mockEvent)
+    })
+
+    assert(mockWinEvents.listeners.length > 1, 'Callbacks were added to window')
+    mockWinEvents.listeners.forEach((listener) => {
+      listener(mockEvent)
+    })
+
+    const entities = bitECS.getAllEntities(Engine.instance)
+    const emulatedInputSourceEntity = entities[entities.length - 1] as Entity
+
+    assert(entityExists(emulatedInputSourceEntity))
+    assert(hasComponent(emulatedInputSourceEntity, InputSourceComponent))
+    assert(hasComponent(emulatedInputSourceEntity, NameComponent))
+
+    const pointerStatePos = getState(InputState).pointerState.position
+    assert(pointerStatePos.x !== 0 && pointerStatePos.y !== 0)
+
+    cleanup()
+    assert(mockDomElm.listeners.length === 0, 'Callbacks were removed from canvas')
+    assert(mockDocEvents.listeners.length === 0, 'Callbacks were removed from document')
+    assert(mockWinEvents.listeners.length === 0, 'Callbacks were removed from window')
+  })
+
+  afterEach(() => {
+    return destroyEngine()
+  })
+})
+
+describe('client input system reactor', () => {
+  beforeEach(() => {
+    createEngine()
+  })
+
+  it('test client input system reactor', async () => {
+    const Reactor = SystemDefinitions.get(ClientInputSystem)!.reactor!
+    const { rerender, unmount } = render(<Reactor />)
+    unmount()
+  })
+
+  it('test client input system execute', async () => {
+    const mockXRInputSource = new MockXRInputSource({
+      handedness: 'left',
+      targetRayMode: 'screen',
+      targetRaySpace: new MockXRSpace() as XRSpace,
+      gripSpace: undefined,
+      gamepad: undefined,
+      profiles: ['test'],
+      hand: undefined
+    }) as XRInputSource
+
+    const mockXRFrame = new MockXRFrame()
+    mockXRFrame.pose.transform.position.x = 0.2134
+    mockXRFrame.pose.transform.position.y = 0.3456
+    mockXRFrame.pose.transform.position.z = 0.2345
+    mockXRFrame.pose.transform.orientation.x = 0.4567
+    mockXRFrame.pose.transform.orientation.y = 0.8455
+    mockXRFrame.pose.transform.orientation.z = 0.2454
+    mockXRFrame.pose.transform.orientation.w = 0.2743
+
+    globalThis.ReferenceSpace.origin = new MockXRReferenceSpace()
```

Comment:
we should be overriding globals in `before` hooks, and cleaning up in `after` hooks